presto-main/src/main/java/com/facebook/presto/execution/scheduler/LegacySqlQueryScheduler.java

```diff
@@ -0,0 +1,882 @@
+    {
+        Map<URI, TaskId> bufferLocations = tasks.stream()
+                .collect(toImmutableMap(
+                        task -> getBufferLocation(task, rootBufferId),
+                        RemoteTask::getTaskId));
+        queryStateMachine.updateOutputLocations(bufferLocations, noMoreExchangeLocations);
+    }
+
+    private static URI getBufferLocation(RemoteTask remoteTask, OutputBufferId rootBufferId)
+    {
+        URI location = remoteTask.getTaskStatus().getSelf();
+        return uriBuilderFrom(location).appendPath("results").appendPath(rootBufferId.toString()).build();
+    }
+
+    /**
+     * returns a List of SqlStageExecutionInfos in a postorder representation of the tree
+     */
+    private List<StageExecutionAndScheduler> createStageExecutions(
+            SectionExecutionFactory sectionExecutionFactory,
+            ExchangeLocationsConsumer locationsConsumer,
+            StreamingPlanSection section,
+            Optional<int[]> bucketToPartition,
+            OutputBuffers outputBuffers,
+            RemoteTaskFactory remoteTaskFactory,
+            SplitSourceFactory splitSourceFactory,
+            Session session)
+    {
+        ImmutableList.Builder<StageExecutionAndScheduler> stages = ImmutableList.builder();
+
+        for (StreamingPlanSection childSection : section.getChildren()) {
+            ExchangeLocationsConsumer childLocationsConsumer = (fragmentId, tasks, noMoreExhchangeLocations) -> {};
+            stages.addAll(createStageExecutions(
+                    sectionExecutionFactory,
+                    childLocationsConsumer,
+                    childSection,
+                    Optional.empty(),
+                    createDiscardingOutputBuffers(),
+                    remoteTaskFactory,
+                    splitSourceFactory,
+                    session));
+        }
+        List<StageExecutionAndScheduler> sectionStages =
+                sectionExecutionFactory.createSectionExecutions(
+                        session,
+                        section,
+                        locationsConsumer,
+                        bucketToPartition,
+                        outputBuffers,
+                        summarizeTaskInfo,
+                        remoteTaskFactory,
+                        splitSourceFactory,
+                        0).getSectionStages();
+        stages.addAll(sectionStages);
+
+        return stages.build();
+    }
+
+    public void start()
+    {
+        if (started.compareAndSet(false, true)) {
+            startScheduling();
+        }
+    }
+
+    private void startScheduling()
+    {
+        requireNonNull(stageExecutions);
+        // still scheduling the previous batch of stages
+        if (scheduling.get()) {
+            return;
+        }
+        executor.submit(this::schedule);
+    }
+
+    private void schedule()
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 36. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.